src/generic/dirctrlg.cpp

```diff
@@ -655,7 +719,158 @@ void wxGenericDirCtrl::OnExpandItem(wxTreeEvent &event)
     ExpandDir(parentId);
 }
 
-void wxGenericDirCtrl::OnCollapseItem(wxTreeEvent &event )
+
+// Private: Used to add items to the right-click menu each time it's popped up.
+void wxGenericDirCtrl::AddRightClickMenuItem(const wxString& label, void(wxGenericDirCtrl::*function)(wxCommandEvent &))
+{
+    if (!m_rightClickMenu)
+        return;
+
+    int id = GetAvailableID();
+    m_rightClickMenu->Append(id, label);
+    Bind(wxEVT_MENU, function, this, id);
+}
+
+// Public: Used by the parent to add items to the right-click menu each time it's popped up. The parent
+// will also need to bind a function to the menu item.
+int wxGenericDirCtrl::NewMenuItem(const wxString& label)
+{
+    if (!m_rightClickMenu)
+        return 0;
+
+    int id = GetAvailableID();
+    m_rightClickMenu->Append(id, label);
+    m_rightClickMenu->UpdateUI();
+    return id;
+}
+
+
+// When a directory is right-clicked, we come here and build the wxMenu.
+// Using a copy of the style passed to the constructor, we can add the
+// right items to the menu.
+void wxGenericDirCtrl::HandleDirMenu()
+{
+    m_rightClickMenu = new wxMenu();
+
+    if (m_style & wxDIRCTRL_EDIT_LABELS)
+    {
+        AddRightClickMenuItem("&Rename", &wxGenericDirCtrl::MenuRename);
+    }
+
+    if (m_style & wxDIRCTRL_RCLICK_MENU_SORT_NAME)
+    {
+        AddRightClickMenuItem("Sort by &Name",          &wxGenericDirCtrl::MenuSortAlpha);        // Only directories need the sort options
+        AddRightClickMenuItem("Sort by Na&me reversed", &wxGenericDirCtrl::MenuSortNameReversed);
+    }
+
+    if (m_style & wxDIRCTRL_RCLICK_MENU_SORT_DATE)
+    {
+        AddRightClickMenuItem("Sort by &Date",         &wxGenericDirCtrl::MenuSortDate);
+        AddRightClickMenuItem("Sort by Da&te reverse", &wxGenericDirCtrl::MenuSortDateReverse);
+    }
+
+    wxCommandEvent event2(wxEVT_DIRCTRL_MENU_POPPED_UP, wxID_MENU_DIR);
+    event2.SetString("Directory Menu");
+    GetEventHandler()->SafelyProcessEvent(event2);              // Create an event so that the parent is informed that the menu has
+                                                                // opened. Then the parent can add menu items if they want.
+    PopupMenu(m_rightClickMenu);
+
+    delete m_rightClickMenu;
```

Comment:
There is a big problem here with not resetting `m_rightClickMenu` to `NULL`.

To resolve it and also to avoid creating the menu on the heap unnecessarily, I'd write this function like this:

    void HandleDirMenu()
    {
        wxMenu menu;
        m_rightClickMenu = &menu;
        wxON_BLOCK_EXIT_NULL(m_rightClickMenu);

        ...
    }

You need to include `wx/scopeguard.h` for the macro above.